Skip superfluous store operations while constructing tensors from data

cbalioglu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #65331

Differential Revision: [D31055036](https://our.internmc.facebook.com/intern/diff/D31055036/)

This is a performance optimization for the use case:

```
tensor = torch.tensor(<large_data>, device='meta')
```

where the current implementation requires a superfluous copy operation even though the target device is a meta.

cbalioglu committed:
Skip superfluous store operations while constructing tensors from data



Differential Revision: [D31055036](https://our.internmc.facebook.com/intern/diff/D31055036/)

[ghstack-poisoned]

cbalioglu committed:
Update on "Skip superfluous store operations while constructing tensors from data"


Differential Revision: [D31055036](https://our.internmc.facebook.com/intern/diff/D31055036/)

This is a performance optimization for the use case:

```
tensor = torch.tensor(<large_data>, device='meta')
```

where the current implementation requires a superfluous copy operation even though the target device is a meta.

[ghstack-poisoned]

ezyang commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -279,11 +281,14 @@ Tensor internal_new_from_data(
+      recursive_store(
+          (char*)tensor.data_ptr(), tensor.sizes(), tensor.strides(), 0,
+          inferred_scalar_type, tensor.dtype().itemsize(), data);
+    }
```
When the source is meta, I'd prefer if we error in this case (because the target will be uninitialized garbage if you don't copy here)

ezyang reviewed (changes requested):
thanks, much appreciated. Just a small semantics change and this is good to go. A test for the case I mention would also be great.

cbalioglu committed:
Update on "Skip superfluous store operations while constructing tensors from data"


Differential Revision: [D31055036](https://our.internmc.facebook.com/intern/diff/D31055036/)

This is a performance optimization for the use case:

```
tensor = torch.tensor(<large_data>, device='meta')
```

where the current implementation requires a superfluous copy operation even though the target device is a meta.

[ghstack-poisoned]

cbalioglu committed:
Update on "Skip superfluous store operations while constructing tensors from data"


Differential Revision: [D31055036](https://our.internmc.facebook.com/intern/diff/D31055036/)

This is a performance optimization for the use case:

```
tensor = torch.tensor(<large_data>, device='meta')
```

where the current implementation requires a superfluous copy operation even though the target device is a meta.

[ghstack-poisoned]

cbalioglu committed:
Update on "Skip superfluous store operations while constructing tensors from data"


Differential Revision: [D31055036](https://our.internmc.facebook.com/intern/diff/D31055036/)

This is a performance optimization for the use case:

```
tensor = torch.tensor(<large_data>, device='meta')
```

where the current implementation requires a superfluous copy operation even though the target device is a meta.

[ghstack-poisoned]

cbalioglu committed:
Update on "Skip superfluous store operations while constructing tensors from data"


Differential Revision: [D31055036](https://our.internmc.facebook.com/intern/diff/D31055036/)

This is a performance optimization for the use case:

```
tensor = torch.tensor(<large_data>, device='meta')
```

where the current implementation requires a superfluous copy operation even though the target device is a meta.

[ghstack-poisoned]

cbalioglu commented:
@ezyang I actually simplified the logic a bit in the second revision which is now also covered by the existing tests.

cbalioglu committed:
Update on "Skip superfluous store operations while constructing tensors from data"


Differential Revision: [D31055036](https://our.internmc.facebook.com/intern/diff/D31055036/)

This is a performance optimization for the use case:

```
tensor = torch.tensor(<large_data>, device='meta')
```

where the current implementation requires a superfluous copy operation even though the target device is a meta.

[ghstack-poisoned]
